web3.js/src/connection.ts

```diff
@@ -4263,75 +4207,273 @@ export class Connection {
+                    ...subscription,
+                    state: 'subscribed',
+                  };
+                  await this._updateSubscriptions();
+                  return;
+                }
+                this._subscriptionsByHash[hash] = {
+                  ...subscription,
+                  state: 'unsubscribed',
+                };
+                await this._updateSubscriptions();
+              })();
+            }
+            break;
+          case 'subscribing':
+          case 'unsubscribing':
+            break;
+        }
+      }),
+    );
+  }
 
-    for (let clientSubscriptionId of rootClientSubscriptionIds) {
-      const sub = this._rootSubscriptions[clientSubscriptionId];
-      this._subscribe(sub, 'rootSubscribe', []);
+  /**
+   *
+   * @internal
+   */
+  _publishNotification<TCallback extends SubscriptionConfig['callback']>(
+    serverSubscriptionId: ServerSubscriptionId,
+    publishArgs: Parameters<TCallback>,
+  ): void {
+    const callbacks =
+      this._subscriptionCallbacksByServerSubscriptionId[serverSubscriptionId];
+    if (callbacks == null) {
+      return;
     }
-
-    for (let clientSubscriptionId of logsClientSubscriptionIds) {
-      const sub = this._logsSubscriptions[clientSubscriptionId];
-      let filter;
-      if (typeof sub.filter === 'object') {
-        filter = {mentions: [sub.filter.toString()]};
-      } else {
-        filter = sub.filter;
+    callbacks.forEach(cb => {
+      try {
+        cb(...(publishArgs as [any, any]));
+      } catch (e) {
+        console.error(e);
       }
-      this._subscribe(
-        sub,
-        'logsSubscribe',
-        this._buildArgs([filter], sub.commitment),
-      );
-    }
+    });
   }
 
   /**
    * @internal
    */
   _wsOnAccountNotification(notification: object) {
-    const res = create(notification, AccountNotificationResult);
-    for (const sub of Object.values(this._accountChangeSubscriptions)) {
-      if (sub.subscriptionId === res.subscription) {
-        sub.callback(res.result.value, res.result.context);
+    const {result, subscription} = create(
+      notification,
+      AccountNotificationResult,
+    );
+    this._publishNotification<AccountChangeCallback>(subscription, [
+      result.value,
+      result.context,
+    ]);
+  }
+
+  /**
+   * @private
+   */
+  _makeSubscription(
+    // The subscription config, before applying defaults
+    rawConfig: SubscriptionConfig,
+  ): ClientSubscriptionId {
+    const clientSubscriptionId = this._nextClientSubscriptionId++;
+    let subscriptionConfig: SubscriptionConfig;
+    // Apply defaults.
+    switch (rawConfig.method) {
+      case 'accountSubscribe':
+      case 'logsSubscribe':
+      case 'programSubscribe':
+        if (rawConfig.params.commitment === undefined) {
+          subscriptionConfig = {
+            ...rawConfig,
+            params: {
+              ...rawConfig.params,
+              commitment: 'finalized',
```

Comment:
Oh, interesting.

I think that in all cases, no matter what the `Connection's` default commitment is, if you make a subscription _without_ supplying a commitment, then the server defaults to `Finalized` (see `unwrap_or_default()` here):

https://github.com/solana-labs/solana/blob/3114c199bde434b47f255bb6fdd6492836fd9a45/rpc/src/rpc_pubsub.rs#L410-L423

https://github.com/solana-labs/solana/blob/b8837c04ec3976c9c16d028fbee86f87823fb97f/sdk/src/commitment_config.rs#L197-L201

The aim of this code here is to apply the _same_ default that the server would apply in the absence of a `commitment`, so that these client configs hash to the same ‘subscription topic’:

```
// Config A
{
  /* ... */
  "params": [
    "83astBRguLMdt2h5U1Tpdq5tjFoJ6noeGwaY3mDLVcri",
    {"commitment": "finalized"}
  ]
}
// Config B
{
  /* ... */
  "params": [
    "83astBRguLMdt2h5U1Tpdq5tjFoJ6noeGwaY3mDLVcri",
    {}
  ]
}
```

I will extra super double check this though.

Reply:
Mmm. Now looking closer at `_buildArgs` I see what you mean.